[Zhang Shi Chen] iP

SkyBlaise99 opened the pull request:
# CS2103T Individual Project

### Project Details

* Application name: ~~Duke~~ __Sora__
* Author: Zhang Shi Chen

### Current features available:

1. list all tasks
1. add a task
1. delete a task
1. mark a task as done
1. find task(s) by keyword

### Todo list:

- [x] Basic Functions
- [x] Unit Testing
- [x] Gradle Support
- [x] GUI

## 

### You can find my codes [here](https://github.com/SkyBlaise99/ip)

This is a sneak peek of my `main` 😄 
``` java
public static void main(String[] args) {
    new Duke("data.txt").run();
}
```

## 

> _"Don't write better error messages, write code that doesn't need them."_ - Jason C. McDonald ([source](https://dev.to/noddydev/don-t-write-better-error-messages-write-code-that-doesn-t-need-them-jason-c-mcdonald-4bko))

Commit:
Add Gradle support

SkyBlaise99 committed:
Level 1 completed

SkyBlaise99 committed:
Level 2 completed

SkyBlaise99 committed:
Level 3 completed

SkyBlaise99 committed:
Level 4 completed

SkyBlaise99 committed:
Updated gitignore

SkyBlaise99 committed:
Updated gitignore v2

SkyBlaise99 committed:
Auto text testing completed

SkyBlaise99 committed:
Level 5 completed

SkyBlaise99 committed:
Updated gitignore v3

SkyBlaise99 committed:
Level 6 completed

SkyBlaise99 committed:
Fix some java conventions

SkyBlaise99 committed:
Update .gitignore to ignore text-ui-test folder

SkyBlaise99 committed:
Stop tracking files in text-ui-test folder

SkyBlaise99 committed:
Delete text-ui-test directory

SkyBlaise99 committed:
Reset .gitignore

SkyBlaise99 committed:
Remove javadoc alignment as part of java convention

SkyBlaise99 committed:
Fix int overflow bug in taskDone and taskDelete method

SkyBlaise99 committed:
Fix all java conventions accordingly

SkyBlaise99 committed:
Complete level 7

Duke: implement storage function, saveData() and loadData()
DukeException: add new DataIntegrityException
Test: update test cases to include testing for saveData() and loadData()

SkyBlaise99 committed:
Complete level 8 (Dates and Times)

Deadline: change time(String) to dateTime(LocalDateTime)
Event: separate time(String) to date(LocalDate) and
startTime(LocalTime) & endTime(LocalTime)
Duke: update addDeadline() & addEvent() accordingly
Test: update test cases to include testing for new date time format

SkyBlaise99 committed:
Add a check for start time later than end time for event

Add test case for the above situation

SkyBlaise99 committed:
Merge branch 'branch-Level-7'

SkyBlaise99 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java
#	text-ui-test/EXPECTED.TXT
#	text-ui-test/input.txt

Update data.txt to match new time display due to level 8

SkyBlaise99 committed:
Update javadocs for Task, Todo, Event and Deadline

SkyBlaise99 committed:
Complete A-MoreOOP

Extract Ui, Storage, Parser, TaskList from Duke
Add Javadoc for the 4 extracted classes
Update Javadoc for Task and its children classes

SkyBlaise99 committed:
Complete A-Packages

* duke for Duke
* duke.exception for all exception classes
* duke.task for all task related classes
* duke.util for Ui, Parser and Storage

SkyBlaise99 committed:
Fix runtest.bat to handle packages

SkyBlaise99 committed:
Change name from setDone() back to markAsDone()

I believe this is not a setter but rather a method, setter should have
a input value.

SkyBlaise99 committed:
Add JUnit for classes in package duke.exception

SkyBlaise99 committed:
Add newline for EOF

SkyBlaise99 committed:
Add JUnit for classes in package duke.task
Remove an extra line in TaskList

SkyBlaise99 committed:
Add JUnit for Parser.java

SkyBlaise99 committed:
Complete A-Jar

Jar file is in ./out/artifacts/ip_jar/ip.jar

SkyBlaise99 committed:
Complete level 9

SkyBlaise99 committed:
Update JUnit for new find feature

SkyBlaise99 committed:
Complete A-JavaDoc

SkyBlaise99 committed:
Merge branch 'branch-A-JavaDoc'

SkyBlaise99 committed:
Merge branch 'master' into branch-A-CodingStandard

SkyBlaise99 committed:
Complete A-CodingStandard

SkyBlaise99 committed:
Merge branch 'branch-A-CodingStandard'

SkyBlaise99 committed:
Merge branch 'branch-Level-9'

SkyBlaise99 committed:
Update new codes from level 9 merge to match coding standard

SkyBlaise99 committed:
Fix code style

Rename some variables to be more descriptive
Split some chain method calls to multiline

SkyBlaise99 committed:
Merge branch 'master' into add-gradle-support

SkyBlaise99 committed:
Complete A-Gradle

Minor update on .gitignore and build.gradle

SkyBlaise99 committed:
Merge branch 'add-gradle-support'

SkyBlaise99 committed:
Copy checkstyle files from AB3

SkyBlaise99 committed:
Fix most of code style wrt checkstyle

SkyBlaise99 committed:
Allow multiple variable declaration

SkyBlaise99 committed:
Fix code style for Storage

SkyBlaise99 committed:
Fix code style for CheckedFunction

SkyBlaise99 committed:
Allow constant variable to be named with all caps

SkyBlaise99 committed:
Fix remaining code style

SkyBlaise99 committed:
Fix some code style for files in test

SkyBlaise99 committed:
Complete A-CheckStyle

SkyBlaise99 committed:
Merge branch 'branch-A-CheckStyle'

SkyBlaise99 committed:
Add Message Enum

Change TaskList and Ui to use Message
Update on code style for the 2 files also

SkyBlaise99 committed:
Update all files according to new code style

Some files are not flagged by checkstyle and hence not updated

SkyBlaise99 committed:
Extract out as newTask() method

SkyBlaise99 committed:
Add Javadoc for classes in exception folder

SkyBlaise99 committed:
Add Javadoc for classes in task folder

SkyBlaise99 committed:
Complete tutorial part 3

SkyBlaise99 committed:
Complete Level 10

Separate out a Main from Duke, now Main runs GUI and Duke run CLI
Shift the 2 controller java files into duke.controller package
Remove temporary function from Storage

SkyBlaise99 committed:
Clean up the code

Delete the temp Duke() constructor created and reuse the original
constructor

SkyBlaise99 committed:
Change Label to Text to prevent long response from being truncated

SkyBlaise99 committed:
Allow the GUI to persist for 1s before exit

SkyBlaise99 committed:
Merge branch 'branch-Level-10'

SkyBlaise99 committed:
Update Javadoc for classes in duke.util

SkyBlaise99 committed:
Change Duke into Sora

Just realized till now all my names are still Duke although I'm calling
it Sora

SkyBlaise99 committed:
Reformat all files

Correct any possible violation of code style due to renaming it to Sora

SkyBlaise99 committed:
Add 'help' function

Update
1. UnknownCommandException: display extra line about 'help' function
2. Message: add help message
3. Test: update tests accordingly

SkyBlaise99 committed:
Add logo for Sora

SkyBlaise99 committed:
Update visual

Changes:
1. dark mode
2. profile pictures
3. small offsets for text and pictures

SkyBlaise99 committed:
Add CSS for ScollPane to fit into dark theme

SkyBlaise99 committed:
Minor edit to visual

1. Remove the title bar
2. Make the profile picture smaller
3. Add 'bye' to supported features

SkyBlaise99 committed:
Improve 'find' function

Find can now search by type of tasks

SkyBlaise99 committed:
Update javadoc and patch the test cases

SkyBlaise99 committed:
Fix shadowJar issue

SkyBlaise99 committed:
Fix code style

ruthpohrp commented on src/main/java/sora/task/Task.java:
```diff
@@ -0,0 +1,63 @@
+     *
+     * @param keyword keyword to match
+     */
+    public boolean match(String keyword) {
```
Maybe put matchesKeyword(...) or hasKeyword(...) or containsKeyword(...) instead?

SkyBlaise99 replied:
Alright updated.

ruthpohrp commented on src/main/java/sora/task/Task.java:
```diff
@@ -0,0 +1,63 @@
+        return matchDescription(keyword) || matchType(keyword);
+    }
+
+    private boolean matchType(String type) {
```
Same as before, may want to put matchesType(...) or isType(...) instead?

SkyBlaise99 replied:
Updated too.

ruthpohrp commented on src/main/java/sora/task/Task.java:
```diff
@@ -0,0 +1,63 @@
+     * Checks if description contains the keyword. Also checks if keyword is a type of Task, e.g. Event, and
+     * in those cases, return all tasks of that type.
+     *
+     * @param keyword keyword to match
```
Remember to add a @return haha

SkyBlaise99 replied:
Oh no... Perhaps I should still use `generate Javadoc` rather than typing myself 🤣

ruthpohrp commented on src/main/java/sora/task/Task.java:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+
+    private boolean matchDescription(String keyword) {
```
Same as above, may want to change to matchesDescription(....) or hasDescription(...)?

SkyBlaise99 replied:
Yup done

ruthpohrp commented on src/main/java/sora/task/TaskList.java:
```diff
@@ -0,0 +1,286 @@
+        return Message.DELETE + "\n  "
+            + task + "\n" + (
+            size == 0
+            ? "You have no more task in the list!"
```
Small grammar error "You have no more task**s** in the list!", really small though

SkyBlaise99 replied:
Thx for pointing out... too long nvr write proper eng alr

greyguy21 commented on src/main/java/sora/Sora.java:
```diff
@@ -0,0 +1,107 @@
+        String input;
+
+        // Program exits only if user inputs "bye"
+        while (!(input = ui.readCommand()).equals("bye")) {
```
I think it would be good to have an intermediate step for reading and assigning the command to the input variable.

ruthpohrp commented on src/main/java/sora/task/TaskList.java:
```diff
@@ -0,0 +1,286 @@
+     * @throws IllegalFormatException if user gives empty or invalid task number
+     * @throws TaskNotFoundException if the task specified by the task number does not exist
+     */
+    public String taskDone(String command) throws IllegalFormatException, TaskNotFoundException {
```
Method 'taskDone' sounds a bit like a boolean method, may want to call method setTaskDone(....) or markTaskDone(...) or something to that effect?

SkyBlaise99 replied:
Not exactly a boolean method, the part where it is actually the boolean method is `task.markAsDone();`
I put `taskDone` since it goes with `taskDelete`

greyguy21 commented on src/main/java/sora/exception/DataIntegrityException.java:
```diff
@@ -0,0 +1,12 @@
+ *
+ * @author Zhang Shi Chen
+ */
+public class DataIntegrityException extends SoraException {
```
I think it is great to have custom exceptions and they are named well too.

ruthpohrp commented on src/main/java/sora/task/TaskList.java:
```diff
@@ -0,0 +1,286 @@
+        String[] info = command.substring(6).split("/at|/from|/to");
+        info = Arrays.stream(info)
+            .map(String::trim)
+            .toArray(String[]::new);
```
I'm not very sure about this, but I think need to add 4 space indentation to this line?

SkyBlaise99 replied:
I'm not really sure about the indentation rule here. Setting it to 8 violates checkstyle rules so I had to switch everything down to 4 which somehow checkstyle is not giving me any troubles. Find it really hard to configure the specific rules so I'm being kinda lazy there.

SkyBlaise99 replied:
Anyways i went to fix it, add some stuffs to suppress warnings haha

ruthpohrp commented on src/main/java/sora/task/TaskList.java:
```diff
@@ -0,0 +1,286 @@
+        String[] info = command.substring(9).split("/by");
+        info = Arrays.stream(info)
+            .map(String::trim)
+            .toArray(String[]::new);
```
Same as above, may need to add 4 space indentation(?)
Happens quite a few times below as well so I'll just comment here 👍

ruthpohrp commented on src/main/java/sora/task/Task.java:
```diff
@@ -0,0 +1,63 @@
+    public String toString() {
+        return "[" + (isDone ? "X" : " ") + "] " + description;
+    }
+}
```
If you have time, may want to add Javadocs for these methods ^-^

SkyBlaise99 replied:
I actually thought that they do not require a javadoc like getter and setters... added them in now.

ruthpohrp commented on src/main/java/sora/task/TaskList.java:
```diff
@@ -0,0 +1,286 @@
+    /**
+     * Deletes the task with the same task number specified by user.
+     *
+     * @param command command entered by user (delete [task number])
```
For all your Javadocs, may want to capitalise the first word to differentiate between parameter and description, e.g. 
```javascript
/**
* ....
* @param command Command entered by user (delete [task number])
* ....
*/
```
Thanks!

SkyBlaise99 replied:
Thx i think i miss read this one... Long way to go for modification man `checkstyle` dun seem to care about this.

ruthpohrp commented on src/main/java/sora/util/Message.java:
```diff
@@ -0,0 +1,36 @@
+        + "   /  /______   /    /  / /____/ / /_|  |\n"
+        + "  /______   /  /    /  /  _   __/  __   |\n"
+        + " _______/  /  /____/  /  / \\  \\/  /  |  |\n"
+        + "/_________/__________/__/   \\__\\_/   |__|\n"
```
kajdnafldfjnsdf I love that you made your custom logo!!! Duke who amirite

SkyBlaise99 replied:
yeah haha, altho it might not be readable since `/` is just 1 slash but to display `\` it needs 2. It looks like this
![image](https://user-images.githubusercontent.com/69448615/131641463-4482a9af-3184-4069-92ad-0aa8bb7aed04.png)

greyguy21 commented on src/main/java/sora/task/TaskList.java:
```diff
@@ -0,0 +1,286 @@
+        // Return confirmation message
+        return Message.DELETE + "\n  "
+            + task + "\n" + (
+            size == 0
```
I think you can consider using an intermediate variable or method to check the for the size of the of the list & the message to return.

SkyBlaise99 replied:
Sorry but I don't really get what you meant here. Is `size` not an intermediate variable?

ruthpohrp reviewed (commented):
Added comments for PR review. Thanks!

greyguy21 commented on src/main/java/sora/task/TaskList.java:
```diff
@@ -0,0 +1,286 @@
+        }
+
+        // Last task is special as it does not need the '\n'
+        sb.append(size)
```
Maybe can just include a conditional clause for this case within the for loop?

SkyBlaise99 replied:
Ok sure!

SkyBlaise99 committed:
Major update according to PR from ruthpohrp

1. Indentation level should be 8 and not 4
2. Edit Javadocs
3. Edit some boolean method names to be more descriptive

greyguy21 commented on src/main/java/sora/util/Storage.java:
```diff
@@ -0,0 +1,170 @@
+    private Task newEvent(String command) throws DataIntegrityException {
+        // [E][ ] project meeting (at: Aug 6 2021, 2:00 PM - 6:00 PM)
+
+        int index = command.lastIndexOf(" (at: ");
```
Perhaps you can abstract this part & take in the prepositions "by" & "at" as parameters?

SkyBlaise99 replied:
This also, can you elaborate more? Do you mean I take `command.lastIndexOf("at")`?

greyguy21 reviewed (commented):
I think your code is very clean & organized and the comments really helped in understanding of the code.

SkyBlaise99 committed:
Add assertion to Storage

Without assertion, it is possible to encounter the situation of
save(null) which causes Null Pointer Exception.

Add assertions to ensure that the tasks being passed in is not null.

Assertions are used to define assumptions about the program state so
that the runtime can verify them.

SkyBlaise99 committed:
Merge pull request #1 from SkyBlaise99/branch-A-Assertions

A-Assertions

SkyBlaise99 committed:
Extract out extractTaskNumber() method

The codes are repeated in 2 methods, which is inefficient and reduce
readability.

The duplicated codes are extracted out as a private method to improve
readability and code quality.

Other methods can also reuse the extracted method in the future.

SkyBlaise99 committed:
Merge pull request #2 from SkyBlaise99/branch-A-CodeQuality

A-CodeQuality

SkyBlaise99 committed:
Create gradle.yml

SkyBlaise99 committed:
Remove I/O redirection tests

SkyBlaise99 committed:
Merge pull request #4 from SkyBlaise99/branch-A-CI

Complete A-CI

SkyBlaise99 committed:
Update according to PR from greyguy21

Use an if statement to remove the duplicated codes.

SkyBlaise99 committed:
Add sort function

SkyBlaise99 committed:
Update new JUnit tests for sort()

SkyBlaise99 committed:
Add 'sort' to help message

SkyBlaise99 committed:
Merge branch 'branch-C-Sort'

SkyBlaise99 committed:
Update error message of 'find' to be more descriptive

SkyBlaise99 committed:
Update JUnit test cases based on the new error message of 'find'

SkyBlaise99 committed:
Change Task to abstract class

To allow Task.getDateTime() to behave differently according to the type
of Task (Todo, Deadline, Event), Task is changed to an abstract class
which has the abstract method getDateTime().

Also updates relevant test cases.

SkyBlaise99 committed:
Update all headers based on feedback from @cs2103-bot

SkyBlaise99 committed:
Add Ui.png

SkyBlaise99 committed:
Set theme jekyll-theme-midnight

SkyBlaise99 committed:
Add UG and partial content

SkyBlaise99 committed:
Merge branch 'master' of https://github.com/SkyBlaise99/ip

SkyBlaise99 committed:
Add basic info & format for all features

SkyBlaise99 committed:
Add a shortcut list

SkyBlaise99 committed:
Add separator between features

SkyBlaise99 committed:
Replace empty ### with <br/>

SkyBlaise99 committed:
Add more info for list

SkyBlaise99 committed:
Add more info for done and delete

SkyBlaise99 committed:
Add more info for find and sort

SkyBlaise99 committed:
Add things to take note for feature

SkyBlaise99 committed:
Update some info regarding date and time for deadline and event

SkyBlaise99 committed:
Add summary table for commands

SkyBlaise99 committed:
Update notes in find

SkyBlaise99 committed:
Fix an issue where mac users are unable to run the jar file

SkyBlaise99 committed:
Add custom title bar

Allow drag, minimize and close

SkyBlaise99 committed:
Merge branch 'branch-CustomTitleBar'

SkyBlaise99 committed:
Make message final

SkyBlaise99 committed:
Update JavaFX versions and Ui.png

SkyBlaise99 committed:
Update all hyperlinks

SkyBlaise99 committed:
Update README.md
